pkg/operator/target_status_test.go

```diff
@@ -0,0 +1,1532 @@
+		{
+			Name:             "PodMonitoring/gmp-test/prom-example-1/metrics",
+			ActiveTargets:    1,
+			UnhealthyTargets: 0,
+			SampleGroups: []v1.SampleGroup{
+				{
+					SampleTargets: []v1.SampleTarget{
+						{
+							Health: "up",
+							Labels: map[model.LabelName]model.LabelValue{
+								"instance": "a",
+							},
+							LastError:                 lastError("err x"),
+							LastScrapeDurationSeconds: "1.2",
+						},
+					},
+					Count: count(1),
+				},
+			},
+			CollectorsFraction: "1",
+		},
+	}
+	expectStatus(t, "first tick", statusTick1)
+
+	active := &prometheusTargetMap[key].Active[0]
+	active.Health = "down"
+	active.LastError = "err y"
+	active.LastScrapeDuration = 5.4
+	// We didn't tick yet so we don't expect a change yet.
+	expectStatus(t, "first wait", statusTick1)
+
+	// Second tick.
+	timer.tick()
+	statusTick2 := []v1.ScrapeEndpointStatus{
+		{
+			Name:             "PodMonitoring/gmp-test/prom-example-1/metrics",
+			ActiveTargets:    1,
+			UnhealthyTargets: 1,
+			SampleGroups: []v1.SampleGroup{
+				{
+					SampleTargets: []v1.SampleTarget{
+						{
+							Health: "down",
+							Labels: map[model.LabelName]model.LabelValue{
+								"instance": "a",
+							},
+							LastError:                 lastError("err y"),
+							LastScrapeDurationSeconds: "5.4",
+						},
+					},
+					Count: count(1),
+				},
+			},
+			CollectorsFraction: "1",
+		},
+	}
+	expectStatus(t, "second tick", statusTick2)
+
+	active = &prometheusTargetMap[key].Active[0]
+	active.Health = "up"
+	active.LastError = "err z"
+	active.LastScrapeDuration = 8.3
+	// We didn't tick yet so we don't expect a change yet.
+	expectStatus(t, "second wait", statusTick2)
+
+	timer.tick()
+	statusTick3 := []v1.ScrapeEndpointStatus{
+		{
+			Name:             "PodMonitoring/gmp-test/prom-example-1/metrics",
+			ActiveTargets:    1,
+			UnhealthyTargets: 0,
+			SampleGroups: []v1.SampleGroup{
+				{
+					SampleTargets: []v1.SampleTarget{
+						{
+							Health: "up",
+							Labels: map[model.LabelName]model.LabelValue{
+								"instance": "a",
+							},
+							LastError:                 lastError("err z"),
+							LastScrapeDurationSeconds: "8.3",
+						},
+					},
+					Count: count(1),
+				},
+			},
+			CollectorsFraction: "1",
+		},
+	}
+	expectStatus(t, "third tick", statusTick3)
+}
+
+// Tests that for pod, targets are fetched correctly (concurrently).
+func TestFetching(t *testing.T) {
+	ctx := context.Background()
+	logger := testr.New(t)
+	concurrency := uint16(4)
+	opts := Options{
+		ProjectID:             "test-proj",
+		Location:              "test-loc",
+		Cluster:               "test-cluster",
+		TargetPollConcurrency: concurrency,
+	}
+	if err := opts.defaultAndValidate(logger); err != nil {
+		t.Fatal("Invalid options:", err)
+	}
+
+	scheme, err := getScheme()
+	if err != nil {
+		t.Fatal("Unable to get scheme")
+	}
+
+	testData := []uint16{0, 1, 2, concurrency - 1, concurrency, concurrency + 1, concurrency * 3}
```

Comment:
What does the 0 case test?

Reply:
Base case where we have no pods. I figured this is a valid test to ensure that having no pods does not freeze the threads or panic anywhere. To make it more clear, I added a comment on this. Do you have any concerns?